File: Bi-KV/DistributedStorage/PageManager.py
from config import *
import time
import threading

eviction_lock = threading.Lock()
# allocate_lock = threading.Lock()

class PageManager:
    def __init__(self, cache_size, page_size, pm_id = 0):
        self.pm_id = pm_id
        # self.buffer = cache_size*['']
        self.page_size = page_size
        self.num_pages = cache_size // page_size
        # set是不是不合适，换成队列？
        self.free_pages = set(range(self.num_pages))
        self.page_table = {}  # 键为item ID，值为{'pages': set, 'last_accessed': int}
        self.current_time = 0  # 模拟时间戳，用于LRU
        # print(f"初始空闲页数:{len(self.free_pages)}")

    def load_item(self, item_id, list_length):
        # 如果id已经存在则直接返回
        if item_id in self.page_table:
            return self.access_item(item_id), []
        """加载列表，必要时换出旧列表"""
        required_pages = (list_length + self.page_size - 1) // self.page_size
        if required_pages > self.num_pages:
            raise ValueError("列表过大，无法存入缓冲区")

        freed_ids = []

        # eviction_lock.acquire()
        # 检查是否需要换出
        if len(self.free_pages) < required_pages:
            freed_ids = self._perform_eviction(required_pages)

        # 分配新页
        allocated_pages = self._allocate_pages(required_pages)
        # 目前的情况看不加锁也没有问题
        # eviction_lock.release()
        self.page_table[item_id] = {
            'pages': allocated_pages,
            'last_accessed': self.current_time,
            'protected': 0
        }
        # for i in allocated_pages:
        #     for j in range(self._page_to_location(i), self._page_to_location(i)+self.page_size):
        #         self.buffer[j] = item_id
        self.current_time += 1
        return allocated_pages,freed_ids

    def access_item(self, item_id):
        """访问列表，更新最近使用时间，返回页号"""
        if item_id not in self.page_table:
            raise KeyError(f"列表 {item_id} 未加载")
        self.page_table[item_id]['last_accessed'] = self.current_time
        self.current_time += 1
        return self.page_table[item_id]['pages']

    def _perform_eviction(self, required_pages):
        """执行换出操作直到有足够空间"""
        # protected_items = sum(len(info['pages']) for info in self.page_table.values() if info['protected'] > 0)
        # paged_items = sum(len(info['pages']) for info in self.page_table.values())
        # print(f"触发换出操作，当前空闲页数:{len(self.free_pages)} 保护数量:{protected_items} 占用数量：{paged_items} 要求页数:{required_pages}")
        # 按LRU排序现有列表
        lru_entries = sorted(
            self.page_table.items(),
            key=lambda item: item[1]['last_accessed'],
        )

        freed_ids = []
        for item_id, info in lru_entries:
            # 如果保护则继续
            if self.page_table[item_id]['protected']!=0:
                continue
            # 换出该列表
            del self.page_table[item_id]
            self.free_pages.update(info['pages'])
            freed_ids.append(item_id)
            if DEBUG:
                print(f"[[{time.time()}]] 换出列表 {item_id}，释放页数 {len(info['pages'])} 空闲页数:{len(self.free_pages)}")
            if len(self.free_pages) >= required_pages:
                break

        if len(self.free_pages) < required_pages:
            # protected_items = sum(len(info['pages']) for info in self.page_table.values() if info['protected'] > 0)
            # print(f"[[{time.time()}]] 无法换出足够页面，当前页面:{list(self.page_table.keys())} 占用数量:{protected_items}")
            raise RuntimeError(f"Unable to free enough pages, even if all items are swapped out") 
        # 返回换出的页号
        return freed_ids

    def set_protected(self, item_id):
        """保护列表不被换出"""
        if item_id not in self.page_table:
            raise KeyError(f"列表 {item_id} 未加载")
        self.page_table[item_id]['protected'] += 1
        # print(f"[[{time.time()}]] 保护item {item_id} {self.page_table[item_id]['protected']}次")

    def remove_protected(self, item_id):
        """取消保护"""
        if item_id not in self.page_table:
            raise KeyError(f"列表 {item_id} 未加载")
        self.page_table[item_id]['protected'] -= 1
        # print(f"[[{time.time()}]] 取消保护item {item_id} {self.page_table[item_id]['protected']}次")

    def _allocate_pages(self, n):
        """分配指定数量的页（无需连续）"""
        if len(self.free_pages) < n:
            raise RuntimeError("内部错误：分配时页面不足")
        allocated = set()
        for _ in range(n):
            # 取第一个可用页（实现可优化）
            page = next(iter(self.free_pages))
            allocated.add(page)
            self.free_pages.remove(page)
        return list(allocated)
    
    def _page_to_location(self, page):
        """将页号转换为实际位置"""
        return page * self.page_size

    def get_loaded_lists(self):
        """返回当前加载的列表ID"""
        return list(self.page_table.keys())

class MultiPageManager:
    '''多个PageManager组成的KV缓存'''
    def __init__(self, cache_size, page_size, kvcahe_num):
        self.kvcahe_num = kvcahe_num
        self.page_managers = [PageManager(cache_size, page_size, pm_id=i) for i in range(kvcahe_num)]
        self.buffer_size = cache_size
        self.page_size = page_size
        self.num_pages = cache_size // page_size
        # 记录每个pm的缓存列表
        self.cached_ids = [set() for _ in range(kvcahe_num)]

    def load_item(self, item_id, list_length):
        """加载列表，必要时换出旧列表"""
        # NOTE 需要注意的是，应该先查询在不在缓存中，如果在缓存中则直接返回页号
        # 计算最大剩余页数
        max_page_num = max(len(pm.free_pages) for pm in self.page_managers)
        if max_page_num>(self.num_pages/10):
            # 选择剩余页数最多的pm
            target_pm = max(self.page_managers, key=lambda x: len(x.free_pages))
        else:
            # 随机选择一个pm 防止负载不均衡
            target_pm = random.choice(self.page_managers)
        target_pm_id = target_pm.pm_id
        allocated_pages, freed_ids = target_pm.load_item(item_id, list_length)
        # 记录已缓存的列表
        self.cached_ids[target_pm_id].add(item_id)
        # 删除已换出的列表
        for freed_id in freed_ids:
            self.cached_ids[target_pm_id].remove(freed_id)
        return target_pm_id, allocated_pages

    def access_item(self, item_id):
        """访问列表，更新最近使用时间，返回页号，否则返回None"""
        for idx, cached_id in enumerate(self.cached_ids):
            if item_id in cached_id:
                return idx,self.page_managers[idx].access_item(item_id)
        return None, None

# 示例用法
if __name__ == "__main__":
    # 初始化：32元素缓冲区，每页4元素 → 8页
    pm = PageManager(cache_size=32, page_size=4)

    # 加载列表A（5元素 → 2页）
    print(pm.load_item("A", 5))  # 例如输出 {0, 1}
    print("当前加载的列表:", pm.get_loaded_lists())
    # print("当前缓冲区状态:", pm.buffer)


    # 加载列表B（7元素 → 2页）
    print(pm.load_item("B", 7))  # 例如输出 {2, 3}
    print("当前加载的列表:", pm.get_loaded_lists())
    # print("当前缓冲区状态:", pm.buffer)


    # 访问列表A更新LRU时间
    pm.access_item("A")

    # 加载列表C（10元素 → 3页，需换出B）
    print(pm.load_item("C", 20))  # 例如输出 {4, 5, 6}
    print("当前加载的列表:", pm.get_loaded_lists())  # 应包含A和C
    # print("当前缓冲区状态:", pm.buffer)


    # 加载列表C（10元素 → 3页，需换出B）
    print(pm.load_item("D", 30))  # 例如输出 {4, 5, 6}
    print("当前加载的列表:", pm.get_loaded_lists())  # 应包含A和C
    # print("当前缓冲区状态:", pm.buffer)

    #测试 MultiPageManager
    mpm = MultiPageManager(32, 4, 2)
    mpm.load_item("A", 5)
    mpm.load_item("B", 7)
    mpm.load_item("C", 20)
    mpm.load_item("D", 30)
    print(mpm.access_item("A"))
    print(mpm.access_item("B"))
    print(mpm.access_item("C"))
    print(mpm.access_item("D"))

File: Go/CacheCoordinator/PageManager.go
package coordinator

import (
	"container/heap"
	"encoding/json"
	"fmt"
	"log"
	"math/rand"
	"os"
	"sort"
	"strings"
	"sync"
	"time"
)

var (
	evictionLock sync.Mutex // 全局锁，模拟 Python 的 eviction_lock
)

// PageEntry 定义缓存项
type PageEntry struct {
	Pages        []int32 // 分配的页面集合
	LastAccessed int64   // 最近访问时间戳（Unix 纳秒）
	Protected    int     // 保护计数
	Priority     int     // 优先级（越小越优先）
}

// Item 优先队列中的元素
type Item struct {
	id           int32   // 缓存项ID
	priority     int     // 优先级
	lastAccessed int64   // 最近访问时间
	pages        []int32 // 页面列表
	index        int     // 堆中的索引
}

// PriorityQueue 优先队列
type PriorityQueue []*Item

func (pq PriorityQueue) Len() int { return len(pq) }

func (pq PriorityQueue) Less(i, j int) bool {
	// 最小堆：优先级小的优先；优先级相同时，时间戳小的优先
	if pq[i].priority == pq[j].priority {
		return pq[i].lastAccessed < pq[j].lastAccessed
	}
	return pq[i].priority < pq[j].priority
}

func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].index = i
	pq[j].index = j
}

func (pq *PriorityQueue) Push(x any) {
	n := len(*pq)
	item := x.(*Item)
	item.index = n
	*pq = append(*pq, item)
}

func (pq *PriorityQueue) Pop() any {
	old := *pq
	n := len(old)
	item := old[n-1]
	old[n-1] = nil
	item.index = -1
	*pq = old[0 : n-1]
	return item
}

// PageManager 单个页面管理器
type PageManager struct {
	pmID      int32
	pageSize  int
	numPages  int
	freePages map[int32]struct{}  // 使用 map 模拟 set
	pageTable map[int32]PageEntry // 缓存项
	mu        sync.Mutex
}

// NewPageManager 初始化 PageManager
func NewPageManager(cacheSize, pageSize int, pmID int32) *PageManager {
	numPages := cacheSize / pageSize
	pm := &PageManager{
		pmID:      pmID,
		pageSize:  pageSize,
		numPages:  numPages,
		freePages: make(map[int32]struct{}),
		pageTable: make(map[int32]PageEntry),
	}
	for i := range numPages {
		pm.freePages[int32(i)] = struct{}{}
	}
	log.Printf("初始空闲页数: %d\n", len(pm.freePages))
	return pm
}

// LoadItem 加载到缓存
func (pm *PageManager) LoadItem(itemID int32, itemLength int, weight int) ([]int32, []int32, time.Duration, time.Duration) {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	// 如果已存在，直接返回
	if _, ok := pm.pageTable[itemID]; ok {
		pages := pm.AccessItem(itemID)
		return pages, []int32{}, 0, 0
	}

	requiredPages := (itemLength + pm.pageSize - 1) / pm.pageSize
	if requiredPages > pm.numPages {
		panic(fmt.Sprintf("列表过大，无法存入缓冲区: %d > %d", requiredPages, pm.numPages))
	}
	var evtCost time.Duration
	var freedIDs []int32
	evictionLock.Lock()
	if len(pm.freePages) < requiredPages {
		start := time.Now()
		freedIDs = pm.performEviction(requiredPages)
		evtCost = time.Since(start)
	}
	start := time.Now()
	allocatedPages := pm.allocatePages(requiredPages)
	allCost := time.Since(start)
	evictionLock.Unlock()

	priority := 0
	if weight != 0 {
		// 用户初始优先级为2
		priority = 2
	}
	// 插入新项
	pm.pageTable[itemID] = PageEntry{
		Pages:        allocatedPages,
		LastAccessed: time.Now().UnixNano(),
		Protected:    0,
		Priority:     priority,
	}
	return allocatedPages, freedIDs, evtCost, allCost
}

// AccessItem 访问列表并更新时间戳
func (pm *PageManager) AccessItem(itemID int32) []int32 {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	if entry, ok := pm.pageTable[itemID]; ok {
		entry.LastAccessed = time.Now().UnixNano()
		pm.pageTable[itemID] = entry
		return entry.Pages
	}
	panic(fmt.Sprintf("列表 %d 未加载", itemID))
}

// performEviction 执行优先队列换出
func (pm *PageManager) performEviction(requiredPages int) []int32 {
	pq := make(PriorityQueue, 0)
	heap.Init(&pq)

	// 将所有未保护的项加入优先队列
	for id, entry := range pm.pageTable {
		if entry.Protected != 0 {
			continue
		}
		if entry.Priority == 2 {
			continue
		}
		heap.Push(&pq, &Item{
			id:           id,
			priority:     entry.Priority,
			lastAccessed: entry.LastAccessed,
			pages:        entry.Pages,
		})
	}

	var freedIDs []int32
	for pq.Len() > 0 && len(pm.freePages) < requiredPages {
		item := heap.Pop(&pq).(*Item)
		delete(pm.pageTable, item.id)
		for _, page := range item.pages {
			pm.freePages[page] = struct{}{}
		}
		freedIDs = append(freedIDs, item.id)
		
		// log.Printf("[[%d]] 换出列表 %d，优先级 %d，释放页数 %d 空闲页数: %d\n",
		// 	time.Now().Unix(), item.id, item.priority, len(item.pages), len(pm.freePages))
		
	}

	if len(pm.freePages) < requiredPages {
		panic(fmt.Sprintf("无法换出足够页面，当前空闲页数: %d，要求页数: %d", len(pm.freePages), requiredPages))
	}
	return freedIDs
}

// SetProtected 设置保护状态
func (pm *PageManager) SetProtected(itemID int32) {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	if entry, ok := pm.pageTable[itemID]; ok {
		entry.Protected++
		pm.pageTable[itemID] = entry
		if DEBUG {
			log.Printf("[[%d]] 保护item %d %d次\n", time.Now().Unix(), itemID, entry.Protected)
		}
	} else {
		panic(fmt.Sprintf("列表 %d 未加载", itemID))
	}
}

// RemoveProtected 取消保护
func (pm *PageManager) RemoveProtected(itemID int32) {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	if entry, ok := pm.pageTable[itemID]; ok {
		entry.Protected--
		pm.pageTable[itemID] = entry
		if DEBUG {
			log.Printf("[[%d]] 取消保护item %d %d次\n", time.Now().Unix(), itemID, entry.Protected)
		}
	} else {
		panic(fmt.Sprintf("列表 %d 未加载", itemID))
	}
}

// allocatePages 分配页面
func (pm *PageManager) allocatePages(n int) []int32 {
	if len(pm.freePages) < n {
		panic(fmt.Sprintf("内部错误：分配时页面不足，剩余: %d，要求: %d", len(pm.freePages), n))
	}
	allocated := make(map[int32]struct{})
	for range n {
		for page := range pm.freePages {
			allocated[page] = struct{}{}
			delete(pm.freePages, page)
			break
		}
	}
	return setToList(allocated)
}

// GetLoadedLists 获取当前加载的列表 ID
func (pm *PageManager) GetLoadedLists() []int32 {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	var ids []int32
	for id := range pm.pageTable {
		ids = append(ids, id)
	}
	return ids
}

func (pm *PageManager) ShowFreePages() {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	log.Printf("[PM %d] 空闲页数: %d\n", pm.pmID, len(pm.freePages))
}

// MultiPageManager 多个 PageManager 的集合
type MultiPageManager struct {
	kvcacheNum   int
	pageManagers []*PageManager
	bufferSize   int
	pageSize     int
	numPages     int
	p1MaxPages   int
	cachedIDs    []map[int32]struct{} // 每个 PM 的缓存 ID 集合
	p2Items      map[int32]struct{}   // 优先级为2的缓存项
	loadDuration time.Duration
	evtDuration  time.Duration
	allDuration  time.Duration
	gcInterval   time.Duration
	ttlInterval  time.Duration
	mu           sync.Mutex
	userCacheCount int 
}

// NewMultiPageManager 初始化 MultiPageManager
func NewMultiPageManager(cacheSize, pageSize, kvcacheNum, gcInterval, ttlInterval int) *MultiPageManager {
	mpm := &MultiPageManager{
		kvcacheNum:   kvcacheNum,
		pageManagers: make([]*PageManager, kvcacheNum),
		bufferSize:   cacheSize,
		pageSize:     pageSize,
		numPages:     cacheSize / pageSize,
		p1MaxPages:   cacheSize / pageSize / 4,
		cachedIDs:    make([]map[int32]struct{}, kvcacheNum),
		p2Items:      make(map[int32]struct{}),
		gcInterval:   time.Duration(gcInterval) * time.Second,
		ttlInterval:  time.Duration(ttlInterval) * time.Second,
		loadDuration: 0,
		evtDuration:  0,
		allDuration:  0,
		userCacheCount: 0, // 【新增】用户类型缓存计数
	}
	for i := range kvcacheNum {
		mpm.pageManagers[i] = NewPageManager(cacheSize, pageSize, int32(i))
		mpm.cachedIDs[i] = make(map[int32]struct{})
	}
	go mpm.ttlGc()
	return mpm
}

// LoadItem 加载到缓存
func (mpm *MultiPageManager) LoadItem(itemID int32, itemLength int, weight int32, itemType string) (int32, []int32) {
	mpm.mu.Lock()
	defer mpm.mu.Unlock()
	// 检查是否已在缓存中
	for idx, cached := range mpm.cachedIDs {
		if _, ok := cached[itemID]; ok {
			pages := mpm.pageManagers[idx].AccessItem(itemID)
			return int32(idx), pages
		}
	}

	// 选择目标 PM
	maxPageNum := 0
	for _, pm := range mpm.pageManagers {
		pm.mu.Lock()
		if len(pm.freePages) > maxPageNum {
			maxPageNum = len(pm.freePages)
		}
		pm.mu.Unlock()
	}

	var targetPM *PageManager
	if maxPageNum > mpm.numPages/10 {
		targetPM = mpm.pageManagers[0] // 默认第一个，后面更新
		for _, pm := range mpm.pageManagers {
			pm.mu.Lock()
			if len(pm.freePages) == maxPageNum {
				targetPM = pm
			}
			pm.mu.Unlock()
		}
	} else {
		targetPM = mpm.pageManagers[rand.Intn(mpm.kvcacheNum)] // 随机选择
	}

	targetPMID := targetPM.pmID
	start := time.Now()
	allocatedPages, freedIDs, evtCost, allCost := targetPM.LoadItem(itemID, itemLength, int(weight))
	mpm.loadDuration += time.Since(start)
	mpm.evtDuration += evtCost
	mpm.allDuration += allCost
	if itemType == "user cache" {
		mpm.p2Items[itemID] = struct{}{}
	}
	mpm.cachedIDs[targetPMID][itemID] = struct{}{}
	for _, freedID := range freedIDs {
		delete(mpm.cachedIDs[targetPMID], freedID)
	}
	mpm.checkAndAdjustP1Pages()
	return targetPMID, allocatedPages
}

func (mpm *MultiPageManager) LoadPreparedItem(itemID int32, itemLength int, weight int32, itemType string) {
	mpm.mu.Lock()
	defer mpm.mu.Unlock()
	// 检查是否已在缓存中
	for idx, cached := range mpm.cachedIDs {
		if _, ok := cached[itemID]; ok {
			mpm.pageManagers[idx].AccessItem(itemID)
			return
		}
	}

	// 选择目标 PM
	maxPageNum := 0
	for _, pm := range mpm.pageManagers {
		pm.mu.Lock()
		if len(pm.freePages) > maxPageNum {
			maxPageNum = len(pm.freePages)
		}
		pm.mu.Unlock()
	}

	var targetPM *PageManager
	if maxPageNum > mpm.numPages/10 {
		targetPM = mpm.pageManagers[0] // 默认第一个，后面更新
		for _, pm := range mpm.pageManagers {
			pm.mu.Lock()
			if len(pm.freePages) == maxPageNum {
				targetPM = pm
			}
			pm.mu.Unlock()
		}
	} else {
		targetPM = mpm.pageManagers[rand.Intn(mpm.kvcacheNum)] // 随机选择
	}

	targetPMID := targetPM.pmID
	start := time.Now()
	_, freedIDs, evtCost, allCost := targetPM.LoadItem(itemID, itemLength, int(weight))
	mpm.loadDuration += time.Since(start)
	mpm.evtDuration += evtCost
	mpm.allDuration += allCost
	mpm.cachedIDs[targetPMID][itemID] = struct{}{}
	for _, freedID := range freedIDs {
		delete(mpm.cachedIDs[targetPMID], freedID)
	}
}

// AccessItem 访问缓存中的列表
func (mpm *MultiPageManager) AccessItem(itemID int32) (int32, []int32) {
	mpm.mu.Lock()
	defer mpm.mu.Unlock()

	for idx, cached := range mpm.cachedIDs {
		if _, ok := cached[itemID]; ok {
			pages := mpm.pageManagers[idx].AccessItem(itemID)
			return int32(idx), pages
		}
	}
	return -1, nil // 表示未找到
}

// checkAndAdjustP1Pages 检查并调整优先级1的页数
func (mpm *MultiPageManager) checkAndAdjustP1Pages() {
	totalP1Pages := 0
	for _, pm := range mpm.pageManagers {
		pm.mu.Lock()
		for _, entry := range pm.pageTable {
			if entry.Priority == 1 {
				totalP1Pages += len(entry.Pages)
			}
		}
		pm.mu.Unlock()
	}

	if totalP1Pages > mpm.p1MaxPages {
		mpm.downgradeP1Items(totalP1Pages - mpm.p1MaxPages)
	}
}

// downgradeP1Items 将优先级1的项降级为优先级0
func (mpm *MultiPageManager) downgradeP1Items(excessPages int) {
	type p1Item struct {
		pmID         int32
		itemID       int32
		lastAccessed int64
		pages        int
	}
	var p1Items []p1Item

	// 收集所有优先级为1的项
	for idx, pm := range mpm.pageManagers {
		pm.mu.Lock()
		for id, entry := range pm.pageTable {
			if entry.Priority == 1 {
				p1Items = append(p1Items, p1Item{
					pmID:         int32(idx),
					itemID:       id,
					lastAccessed: entry.LastAccessed,
					pages:        len(entry.Pages),
				})
			}
		}
		pm.mu.Unlock()
	}

	// 按 lastAccessed 升序排序（LRU）
	sort.Slice(p1Items, func(i, j int) bool {
		return p1Items[i].lastAccessed < p1Items[j].lastAccessed
	})

	// 降级项直到释放足够的页数
	releasedPages := 0
	for _, item := range p1Items {
		if releasedPages >= excessPages {
			break
		}
		pm := mpm.pageManagers[item.pmID]
		pm.mu.Lock()
		if entry, ok := pm.pageTable[item.itemID]; ok && entry.Priority == 1 {
			entry.Priority = 0
			pm.pageTable[item.itemID] = entry
			releasedPages += item.pages
			if DEBUG {
				log.Printf("[[%d]] 降级item %d from priority 1 to 0\n", time.Now().Unix(), item.itemID)
			}
		}
		pm.mu.Unlock()
	}
}

func (mpm *MultiPageManager) ttlGc() {
	ticker := time.NewTicker(mpm.gcInterval)
	defer ticker.Stop()
	for range ticker.C {
		if DEBUG {
			log.Printf("[[%d]] 定期清理过期缓存\n", time.Now().Unix())
		}
		mpm.downgradeP2Items()
		// 每次GC后检查优先级1的容量
		mpm.mu.Lock()
		mpm.checkAndAdjustP1Pages()
		mpm.mu.Unlock()
	}
}

func (mpm *MultiPageManager) downgradeP2Items() {
	mpm.mu.Lock()
	defer mpm.mu.Unlock()
	currentTime := time.Now().UnixNano()
	ttl := mpm.ttlInterval.Nanoseconds()

	for itemID := range mpm.p2Items {
		for _, pm := range mpm.pageManagers {
			pm.mu.Lock()
			if entry, ok := pm.pageTable[itemID]; ok && entry.Priority == 2 {
				elapsed := currentTime - entry.LastAccessed
				if elapsed >= ttl {
					entry.Priority = 1
					pm.pageTable[itemID] = entry
					if DEBUG {
						log.Printf("[[%d]] 降级item %d from priority 2 to 1 due to TTL\n", time.Now().Unix(), itemID)
					}
					delete(mpm.p2Items, itemID)
				}
			}
			pm.mu.Unlock()
		}
	}
}

func (pm *MultiPageManager) ShowFreePages() {
	pm.mu.Lock()
	defer pm.mu.Unlock()
	for _, pm := range pm.pageManagers {
		pm.ShowFreePages()
	}
}

func (pm *MultiPageManager) ReadPreparedData(dataPath string, indexPath string) {
	file, _ := os.ReadFile(dataPath)
	var data map[string]int
	json.Unmarshal(file, &data)
	log.Printf("读取预加载数据: %d 条\n", len(data))
	indexFile, _ := os.ReadFile(indexPath)
	var userIDs []int
	json.Unmarshal(indexFile, &userIDs)
	log.Printf("读取预加载数据热度索引: %d 条\n", len(userIDs))
	maxNumPages := pm.kvcacheNum * pm.numPages
	totalPages := 0
	for _, k := range userIDs {
		var v int
		if strings.Contains(indexPath, "user") {
			v = data[fmt.Sprintf("%d", k+2000000)]
		} else {
			v = data[fmt.Sprintf("%d", k)]
		}
		pageNum := (v + pm.pageSize - 1) / pm.pageSize
		pm.LoadPreparedItem(int32(k), v, 0, "prepared")
		totalPages += pageNum
		if totalPages > maxNumPages {
			log.Printf("预加载数据超过最大页数限制: %d > %d\n", totalPages, maxNumPages)
			break
		}
	}
	pm.ShowFreePages()
}

func (mpm *MultiPageManager) ShowUserCacheCount() {
	log.Printf("当前用户类型缓存数量（ID > 2000000）: %d\n", len(mpm.p2Items))
}

// 调试开关
const DEBUG = false // 可根据需要设置为 true
